### PR TITLE
Resolve segfaults when TF & Catalyst are both imported

### PR DESCRIPTION
**Pull request by dime10**
Global symbols in our shared libraries can conflict with identically named global symbols in other libraries loaded in the same process. This can easily happen when two projects include the same library, such as TensorFlow and Catalyst both including LLVM in their distributions.

All core compiler dependencies are now compiled with default hidden visibility, including LLVM/MLIR, Enzyme, MLIR-HLO, and Catalyst. Since we link dependencies statically into our main `compiler_driver` library this does not pose any issues. It does mean however mean that most symbols will be inaccessible outside of the `compiler_driver` library.

**Update:** Hidden visibility completely removes symbols from stack traces on Linux, because the backtrace mechanism only looks at the dynamic symbol table. For this reason, "protected" visibility is used instead on Linux, which still exports most symbols to the dynamic symbol table but disallows pre-empting internal definitions for those symbols. On macOS, hidden protected visibility does not appear to be supported, but hidden symbols still appear in stack traces.

Note: While the potential for symbol conflicts exists for both Linux & macOS, in my testing the crashes only occurred on macOS.

**Tested:** 
- Linux: https://github.com/PennyLaneAI/catalyst/actions/runs/7630364778
- macOS (arm): https://github.com/PennyLaneAI/catalyst/actions/runs/7631905361
- macOS (x86): currently failing in the base branch

[sc-54428]

### CODE REVIEW

**Review by erick-xanadu — APPROVED** [review]
:partying_face: :raccoon:

**Comment by erick-xanadu** [comment]
@dime10 changelog?

**Comment by erick-xanadu** [comment]
@dime10 you can also remove some infrastructure in pytests that always imports catalyst before tensorflow for the autograph tests.

File: mlir/Makefile [hunk 1]
```diff
@@ -11,6 +11,7 @@ MHLO_BUILD_DIR?=$(MK_DIR)/mlir-hlo/bazel-build
 ENZYME_BUILD_DIR?=$(MK_DIR)/Enzyme/build
 RT_BUILD_DIR?=$(MK_DIR)/../runtime/build
 ENABLE_ASAN?=OFF
+BUILD_TYPE?=Release
```
Comment (maliasadi) [inline mlir/Makefile:14]: good catch!

**Review by maliasadi — APPROVED** [review]
:tada:

**Pull request merged**